python/tutorials/gluon/08-warp-specialization.py

```diff
@@ -0,0 +1,655 @@
+    grid = (triton.cdiv(xnumel, XBLOCK), )
+
+    block_shape = [XBLOCK, YBLOCK]
+    layout = gl.NVMMASharedLayout.get_default_for(block_shape, gl.float32)
+    a_desc = TensorDescriptor(a, a.shape, a.stride(), block_shape, layout)
+    b_desc = TensorDescriptor(b, b.shape, b.stride(), block_shape, layout)
+    c_desc = TensorDescriptor(c, c.shape, c.stride(), block_shape, layout)
+
+    # By default, a warp-specialized kernel assumes maxnreg=256, the maximum
+    # allowed per thread, in order to determine how to reallocate registers.
+    # We need to intentionally set the register limit. Since the kernel will
+    # have `num_warps+4` warps total, register usage will be
+    #
+    #     maxnreg * (num_warps+4) * 32
+    #
+    # Keep this in mind when deciding how much occupancy you want.
+    elementwise_add_warp_specialized_kernel[grid](  #
+        a_desc, b_desc, c_desc, xnumel, ynumel,  #
+        XBLOCK, YBLOCK, num_load_buffers, num_store_buffers,  #
+        num_warps=num_warps, maxnreg=128)
+
+
+@pytest.mark.parametrize("xnumel, ynumel", [(1000, 2000), (4000, 120)])
+@pytest.mark.parametrize("XBLOCK, YBLOCK", [(32, 64)])
+@pytest.mark.parametrize("num_load_buffers, num_store_buffers", [(1, 1), (2, 2)])
+@pytest.mark.parametrize("num_warps", [4, 8])
+@pytest.mark.skipif(not is_hopper_or_newer(), reason="Requires Hopper or newer")
+def test_elementwise_add_warp_specialized(xnumel, ynumel, XBLOCK, YBLOCK, num_load_buffers, num_store_buffers,
+                                          num_warps):
+    a = torch.randn(xnumel, ynumel, device="cuda")
+    b = torch.randn(xnumel, ynumel, device="cuda")
+    c = torch.empty_like(a, device="cuda")
+    elementwise_add_warp_specialized(a, b, c, XBLOCK, YBLOCK, num_load_buffers, num_store_buffers, num_warps)
+    torch.testing.assert_close(a + b, c, atol=0, rtol=0)
+
+
+if __name__ == "__main__":
+    print("Benchmarking elementwise_add")
+    print("============================")
+    xnumel, ynumel = 32 * 1024, 32 * 1024
+    A = torch.randn(xnumel, ynumel, device="cuda")
+    B = torch.randn(xnumel, ynumel, device="cuda")
+    C = torch.empty_like(A, device="cuda")
+
+    XBLOCK = 64
+    YBLOCK = 128
+    num_load_buffers = 3
+    num_store_buffers = 1
+    num_warps = 4
+
+    ms = triton.testing.do_bench(lambda: t4.elementwise_add_tma(  #
+        A, B, C, XBLOCK, YBLOCK, num_load_buffers))
+    print(f"elementwise_add_tma: {ms:.2f} ms")
+
+    ms = triton.testing.do_bench(lambda: elementwise_add_warp_specialized(  #
+        A, B, C, XBLOCK, YBLOCK, num_load_buffers, num_store_buffers, num_warps))
+    print(f"elementwise_add_warp_specialized: {ms:.2f} ms")
+    print()
+
+# %%
+# Results on GB200:
+#
+# ```
+# elementwise_add_tma: 2.02 ms
+# elementwise_add_warp_specialized: 1.96 ms
+# ```
+#
+# The warp specialized implementation ekes out another performance gain over
+# the software pipelined kernel from 04-tma.py by relying on the warp scheduler
+# to hide latencies.
```

Comment:
Worth stating that we're massively bandwidth bound, hence why it's such a modest difference. In fact I wonder if it might be interesting to do something more complex like a classic mandelbrot or similar to better showcase the latency hiding abilities.

Reply:
I added a note about being bandwidth bound. I agree mandelbrot would be a very cool example but don't think I'll have time to implement that :(